warehouse/query-core/src/main/java/datawave/query/transformer/GroupingTransform.java

```diff
@@ -157,17 +158,34 @@ private Key createAttrKey(String fieldName) {
      *            an iterator source
      * @return the flushed value that is an aggregation from the source iterator
      */
-    public Iterator<Entry<Key,Document>> getGroupingIterator(final Iterator<Entry<Key,Document>> in, int max) {
+    public Iterator<Entry<Key,Document>> getGroupingIterator(final Iterator<Entry<Key,Document>> in, int groupFieldsBatchSize, YieldCallback yieldCallback) {
         
         return new Iterator<Entry<Key,Document>>() {
             
             Entry<Key,Document> next;
             
             @Override
             public boolean hasNext() {
-                for (int i = 0; i < max && in.hasNext(); i++) {
-                    Entry<Key,Document> lastEntry = in.next();
-                    GroupingTransform.this.apply(lastEntry);
+                for (int i = 0; i < groupFieldsBatchSize; i++) {
+                    if (in.hasNext()) {
+                        Entry<Key,Document> lastEntry = in.next();
+                        if (lastEntry != null) {
+                            GroupingTransform.this.apply(lastEntry);
+                        } else if (yieldCallback != null && yieldCallback.hasYielded()) {
+                            log.trace("lastEntry is null and yield was called");
+                            yieldCallback.yield(null); // undo the yield
+                        } else {
+                            // lastEntry was null and there was no yield
+                            break;
+                        }
+                    } else if (yieldCallback != null && yieldCallback.hasYielded()) {
+                        // hasNext is false and yield has been called. Undo the yield.
+                        log.trace("hasNext is false because yield was called");
+                        yieldCallback.yield(null);
```

Comment:
Don't we want to break here?

Reply:
By calling yieldCallback.yield(null), the yield is disabled and hasNext() will likely return true when we hit it again at the top of the loop. Glad to see your initial instinct was the same as mine: my initial instinct was to break here, and that is why my early tests today failed to gather and aggregate all of the entries after a yield. Once I allowed the loop to continue here, the same grouping was done whether there were yields or not.